### ISSUE TITLE
React::ServerRendering::PrerenderError within shoryuken background job on Heroku

### ISSUE THREAD

**Issue by jparker10** [reactjs/react-rails#649]
Help us help you! Please choose one:

- [x] My app crashes with `react-rails`, so I've included the stack trace and the exact steps which make it crash.
- [ ] My app doesn't crash, but I'm getting unexpected behavior. So, I've described the unexpected behavior and suggested a new behavior.
- [ ] I'm trying to use `react-rails` with another library, but I'm having trouble. I've described my JavaScript management setup (eg, Sprockets, Webpack...), how I'm trying to use this other library, and why it's not working.
- [ ] I have another issue to discuss.

-------

I can render server side from a Heroku console just fine, but when I try to render the same content via a Heroku worker, I often get this stack track. It's not every time... just occasionally. Would appreciate any insight.

```
Encountered error "ExecJS::ProgramError" when prerendering MailPublish with { SNIPPED }
/app/vendor/bundle/ruby/2.3.0/gems/execjs-2.7.0/lib/execjs/ruby_racer_runtime.rb:36:in `rescue in block in eval'
/app/vendor/bundle/ruby/2.3.0/gems/execjs-2.7.0/lib/execjs/ruby_racer_runtime.rb:33:in `block in eval'
/app/vendor/bundle/ruby/2.3.0/gems/execjs-2.7.0/lib/execjs/ruby_racer_runtime.rb:75:in `block in lock'
/app/vendor/bundle/ruby/2.3.0/gems/execjs-2.7.0/lib/execjs/ruby_racer_runtime.rb:73:in `Locker'
/app/vendor/bundle/ruby/2.3.0/gems/execjs-2.7.0/lib/execjs/ruby_racer_runtime.rb:73:in `lock'
/app/vendor/bundle/ruby/2.3.0/gems/execjs-2.7.0/lib/execjs/ruby_racer_runtime.rb:32:in `eval'
/app/vendor/bundle/ruby/2.3.0/gems/react-rails-1.9.0/lib/react/server_rendering/exec_js_renderer.rb:23:in `render'
/app/vendor/bundle/ruby/2.3.0/gems/react-rails-1.9.0/lib/react/server_rendering/sprockets_renderer.rb:41:in `render'
/app/vendor/bundle/ruby/2.3.0/gems/react-rails-1.9.0/lib/react/server_rendering.rb:17:in `block in render'
/app/vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.1/lib/connection_pool.rb:64:in `block (2 levels) in with'

**Comment by rmosolgo** [comment]
> (Describe your issue here)

😬

**Comment by jparker10** [comment]
Sorry! Copy paste fail. It's now there.

**Comment by jparker10** [comment]
Current feeling is that the Javascript rendering engine process is dying in the background within the Heroku Dyno and the process isn't getting restarted. Any thoughts on how I can test for that condition?

**Comment by rmosolgo** [comment]
I never heard of that before! Generally, ProgramError is used to rescue some error from V8: 

https://github.com/rails/execjs/blob/ff3f0fd993653d3fd742fe2174a8b23baae3cc87/lib/execjs/ruby_racer_runtime.rb#L14

then 

https://github.com/rails/execjs/blob/ff3f0fd993653d3fd742fe2174a8b23baae3cc87/lib/execjs/ruby_racer_runtime.rb#L89

But, we're not showing a good error message! I wonder if there is more info there that we're missing. Want to try this? 

https://github.com/reactjs/react-rails/pull/650

Alternatively, we could add some error handling into the react-rails server JS instead of rescuing from ExecJS only...

**Comment by blueboxjesse** [comment]
Thanks. Gone ahead and implemented 650 and will wait for the error to crop up again. I've got a workaround where I exit the process if we encounter that error. Heroku will then boot the worker again, it will pick up the job a second time and complete it. Not graceful, but at-least it unblocks the criticality of this issue. :)

**Comment by jesseproudman** [comment]
Looks like we're getting the same thing: 

Encountered error "#<ExecJS::ProgramError: ExecJS::ProgramError>" when prerendering MailerBroadcast.

**Comment by jesseproudman** [comment]
@rmosolgo Looks like the issue is still occurring on a consistent basis.

**Comment by rmosolgo** [comment]
I'm bummed that it didn't come with more info :S 


Gem-wise, I'm not exactly sure how to keep digging. You can see in execjs where `ProgramError` is used to rescue from a JavaScript error during `eval`: 

https://github.com/rails/execjs/blob/ff3f0fd993653d3fd742fe2174a8b23baae3cc87/lib/execjs/ruby_racer_runtime.rb#L33-L37

That makes me think that _this_ error is being caused by some JavaScript error while rendering this component. 

Some ways to assess that possibility: 

- Is there anything unusual about the props that cause this error? For example, are any of the values missing, or do any of the values have a different type than usual (eg, string instead of number), or are any of the values `null`? 
- Along the same lines, when you render in a Heroku console, do you use the same props from the error message? 

Also, can you tell me a bit about your server rendering setup? What are your config values? 

Is there any chance that some side-effects from previous renders are interfering with a later render?

**Comment by jesseproudman** [comment]
Going to try miniracer and see if that gets me anywhere.

The JS is fine since I can re-run the exact same code (with the same props) on a heroku console and it works fine.

Using default server rending config. Nothin unique.

> Is there any chance that some side-effects from previous renders are interfering with a later render?

This seems like what could be happening, but I can't replicate the issue in a console... Only seems to happen in the job workers.

**Comment by rmosolgo** [comment]
Another possibility could be something to do with threaded execution, although I _thought_ we were safe from our use of `ConnectionPool`: 

https://github.com/reactjs/react-rails/blob/master/lib/react/server_rendering.rb#L16-L18

**Comment by rmosolgo** [comment]
mini_racer seems like a good thing to try, maybe it will shed some light on the issue!

**Comment by jesseproudman** [comment]
Is there an easy way to disable the connection pooling?

**Comment by rmosolgo** [comment]
Hmm, I guess if the setting is `1`, we could skip it anyways -- it looks like ExecJS has a `lock { ... }` around execution! 

But there's no built-in way for that, you could remove it from `server_rendering.rb` by hand.

**Comment by jesseproudman** [comment]
Ok cool. Will give mini_racer a shot and then try that if I continue to have trouble. Thanks mate!

**Comment by rmosolgo** [comment]
Any news here?

**Comment by jesseproudman** [comment]
Seems to have cleared up the staging site, though email volumes there are very low. Pushing to production today so should know more this week.

**Comment by jesseproudman** [comment]
Ok, so mini_racer gives us a new error. Executing this same code path via the console seems to work just fine. We're only getting this on the second or third executions within the shoryuken  background job.

`Message: <error> at undefined:0:0
(execjs):1 (none)
/app/vendor/bundle/ruby/2.3.0/gems/mini_racer-0.1.7/lib/mini_racer.rb:167 eval_unsafe
/app/vendor/bundle/ruby/2.3.0/gems/mini_racer-0.1.7/lib/mini_racer.rb:167 block (2 levels) in eval
/app/vendor/bundle/ruby/2.3.0/gems/mini_racer-0.1.7/lib/mini_racer.rb:217 timeout
/app/vendor/bundle/ruby/2.3.0/gems/mini_racer-0.1.7/lib/mini_racer.rb:166 block in eval
/app/vendor/bundle/ruby/2.3.0/gems/mini_racer-0.1.7/lib/mini_racer.rb:63 block in with_lock
/app/vendor/bundle/ruby/2.3.0/gems/mini_racer-0.1.7/lib/mini_racer.rb:63 synchronize
/app/vendor/bundle/ruby/2.3.0/gems/mini_racer-0.1.7/lib/mini_racer.rb:63 with_lock
/app/vendor/bundle/ruby/2.3.0/gems/mini_racer-0.1.7/lib/mini_racer.rb:164 eval
/app/vendor/bundle/ruby/2.3.0/gems/execjs-2.7.0/lib/execjs/mini_racer_runtime.rb:27 block in eval
/app/vendor/bundle/ruby/2.3.0/gems/execjs-2.7.0/lib/execjs/mini_racer_runtime.rb:66 translate
/app/vendor/bundle/ruby/2.3.0/gems/execjs-2.7.0/lib/execjs/mini_racer_runtime.rb:26 eval
/app/vendor/bundle/ruby/2.3.0/bundler/gems/react-rails-6ea7524f5770/lib/react/server_rendering/exec_js_renderer.rb:36 render_from_parts
/app/vendor/bundle/ruby/2.3.0/bundler/gems/react-rails-6ea7524f5770/lib/react/server_rendering/exec_js_renderer.rb:17 render
/app/vendor/bundle/ruby/2.3.0/bundler/gems/react-rails-6ea7524f5770/lib/react/server_rendering/sprockets_renderer.rb:38 render
/app/vendor/bundle/ruby/2.3.0/bundler/gems/react-rails-6ea7524f5770/lib/react/server_rendering.rb:17 block in render`

**Comment by jesseproudman** [comment]
Trying no connection pooling now.

**Comment by jesseproudman** [comment]
Getting rid of the connection pooling seems to have resolved this issue for me:
https://github.com/acrelo/react-rails/blob/master/lib/react/server_rendering.rb#L16-L18

**Comment by rmosolgo** [comment]
Glad to hear it! Maybe we can refractor so that when the pool size is 1, it just _doesn't make a pool_. It looks like execjs takes care of wrapping the context in a mutex anyways.

**Comment by jesseproudman** [comment]
Sounds good. LMK how I can assist from here!

**Comment by BookOfGreg** [comment]
Is this still an issue? Going to close for now, though it looks like there is some potential enhancement in not using the ConnectionPool if the size is 1.
